Audio asset library testing

pike983 opened the pull request:
Adding the audio asset library.

Snow106 committed:
Added irrKlang to build

Snow106 committed:
fixed irrKlang .dll dependency

pike983 committed:
Implemented irrklang audio library with a AudioPlayerSingleton.

Created a wrapper interface to use the irrklang libraries most important functions. Uses a singleton to maintain the same audio engine.

Snow106 commented on out/build/x64-Debug/thirdparty/glfw3/src/glfw3.pc:
```diff
@@ -0,0 +1,13 @@
+prefix=C:/University/BTech TERM 3/COMP 8551/EngineProjectRedo/GameEngine/out/install/x64-Debug
```
What's the purpose of these files? where these missing from the previous GLFW3 install?

Snow106 commented on Engine/src/Engine/Audio.h:
```diff
@@ -0,0 +1,86 @@
+
+// A singleton for controlling sounds throughout the application.
+// Contacts: James
+//
```
missing namespace

Snow106 commented on Engine/src/Engine/Audio.h:
```diff
@@ -0,0 +1,86 @@
+#define _AUDIO_H_
+
+#include <iostream>
+#include "..\irrKlang\irrKlang.h"
```
shouldn't use relative paths to include irrKlang. This should already be available from CMake as an additional include

Snow106 commented on Engine/src/Engine/Audio.h:
```diff
@@ -0,0 +1,86 @@
+// For further examples of how to use the ISoundEngine and
+// ISound, view: https://www.ambiera.com/irrklang/docu/index.html
+
+/*
```
Coding standards: don't use /* comments.

Snow106 commented on Engine/src/Engine/Audio.cpp:
```diff
@@ -0,0 +1,69 @@
+	{
+		return theSound;
+	}
+	return NULL;
```
use nullptr

Snow106 commented on Engine/src/Engine/Audio.cpp:
```diff
@@ -0,0 +1,69 @@
+
+AudioPlayerSingleton* AudioPlayerSingleton::getInstance()
+{
+	if (instance == 0)
```
Should check for nullptr instead of 0

Snow106 commented on Engine/src/Engine/Audio.cpp:
```diff
@@ -0,0 +1,69 @@
+ISound* AudioPlayerSingleton::playSound(const char* soundPath, bool loop, bool startPaused, bool useSoundEffects)
+{
+	ISound* theSound = engine->play2D(soundPath, loop, startPaused, true, ESM_AUTO_DETECT, useSoundEffects);
+	if (theSound != 0)
```
does the library actually return a 0 or a nullptr if there's an error? If this is to keep parity with the library then it's probably fine

pike983 replied:
The library returns a 0 if it can't play the sound.

Snow106 commented on Engine/src/Engine/Audio.cpp:
```diff
@@ -0,0 +1,69 @@
+		return theSound;
+	}
+	return NULL;
+	//TODO: implement alternate media folder search if multiple media folders
```
It's probably better to leave the TODO here and create a task in Trello with the relevant code to be added

Snow106 commented on Engine/src/Engine/Audio.cpp:
```diff
@@ -0,0 +1,69 @@
+// using this funtion using the returned ISound*.
+// More information on sound effects: https://www.ambiera.com/irrklang/docu/index.html#effects
+// 
+ISound* AudioPlayerSingleton::play3DSound(const char* soundPath, vec3df sound3DPosition, bool loop, bool beginPaused, bool useSoundEffects)
```
is pass by value ok here? (for the vec3def)

pike983 replied:
Yeah, I think it should be. The returned ISound lets the position be changed later.

Snow106 commented on Engine/src/Engine/Audio.h:
```diff
@@ -0,0 +1,86 @@
+class AudioPlayerSingleton
+{
+private:
+	static AudioPlayerSingleton* instance;
```
guidelines: m_instance

Snow106 reviewed (changes requested):
Everything in out/build/x64-Debug should not be merged into develop.
add relevant filepaths to .gitignore, though I would note that it seems a lot of these were added because certain configurations were changed from VS. Configs in this project should be done through Cmake, reach out if you need help with that

Snow106 committed:
Fixed cmake irrklang include issue

pike983 committed:
Addressing PR comments.

Addressed the pull request comments left by Snow. Added some corrections to the sample code in Audio.h.

pike983 committed:
Merge branch 'develop' into audioAssetLibraryTesting

Snow106 commented on Engine/src/Engine/Audio.h:
```diff
@@ -4,80 +4,83 @@
-{
-private:
-	static AudioPlayerSingleton* instance;
+namespace IAudio {
```
Bracket in new line

Snow106 replied:
also namespace should still be Engine

pike983 committed:
Update Audio.h

Fixed namespace.

pike983 committed:
Update Audio.cpp

Fixing missed namespace update.

mikesinclair66 reviewed (dismissed):
src/Engine/Audio.cpp can't use "using namespace example_name";
guidelines

Snow106 commented:
> src/Engine/Audio.cpp can't use "using namespace example_name"; guidelines

While I personally perfer leaving "using" out of .cpp files, the guidelines specifically allow the use of using in implementation files "Writing using namespace in an implementation file is fine though."
